src/opart_gaussian.c

```diff
@@ -15,21 +31,97 @@ int opart_gaussian(int n_data, double *data_ptr, double penalty, double *cost_pt
 	  return NUM_OF_DATA_VALUES_LESS_THAN_ZERO;
 	}
 
+	int i;
+	int j;
+	int temp;
+	int maxPos;
+
+	//allocate dynamic memory for global variables
+	sums = (double*)malloc((n_data + 1)*(sizeof(double)));
+	sums_sq = (double*)malloc((n_data + 1)*(sizeof(double)));
+	dp = (double*)malloc((n_data + 1)*(sizeof(double)));
+	positions = (int*)malloc((n_data + 1)*(sizeof(int)));
+
+	//assign penalty to beta so that it's available globally
+  beta = penalty;
+
+  //store cumulative sums and squared sums for O(1) access to segment cost
   InitializeSums(data_ptr, sums, n_data);
-	InitializeSumsSq(data_ptr, sq_sums, n_data);
+	InitializeSumsSq(data_ptr, sums_sq, n_data);
 
-	int i;
-	for(i = 0; i < n_data; i++){
-		cost_ptr[i] = data_ptr[i] / penalty;
+	//Compute optimal cost values and segment ends
+  FindOptimalSegments(data_ptr, n_data);
+
+  //Copy the optimal cost values to cost_ptr for return
+	for(i = 1; i <= n_data; i++){
+		cost_ptr[i - 1] = dp[i];
+	}
+
+	//Traceback the optimal segment ends and copy to end_ptr for return
+	end_ptr[0] = n_data;
+	i = n_data;
+	j = 1;
+	while(positions[i] > 1){
+		end_ptr[j++] = positions[i] - 1;
+	  i = positions[i] - 1;
+	}
+	i = 0;
+	maxPos = j;
+	j--;
+	while(j > i){
```

Comment:
what are these while loops doing? using for loops would be better (easier to read/understand)